news/_posts/2021-11-15-form-theme-new-in-1.7.8.md

```diff
@@ -0,0 +1,110 @@
+                'options' => [
+                    'constraints' => [
+                        new Length([
+                            'max' => $maxLength,
+                            'maxMessage' => sprintf('This field cannot be longer than %limit% characters', $maxLength)
+                        ])
+                    ]
+                ]
+            ])
+            ->add('email', EmailType::class, [
+                'label' => $this->trans('Email address', 'Admin.Global'),
+                'required' => false,
+                'constraints' => [
+                    new Email(['message' => '%s is invalid.'])
+                ],
+            ])
+        ;
+```
+
+The above PHP code basically states that this form contains:
+- a 'title' input that has an upper limit length
+- an 'email' input that must contain a valid email
+
+Here we have the structure of our form. However, this does not specify the HTML structure of the form, such as how inputs should be rendered, should they be wrapped in div, what CSS class to apply...
+
+In Symfony, you can define the HTML structure of a form either by writing it down, tag by tag, or you can delegate this responsibility to a Form Theme.
+
+The Form Theme is a set of macros that describes how each form item is rendered into HTML. It will state how one form field, properly specified, should be rendered as an HTML structure. For example, it controls how to render in HTML a 'textarea' field with a validation message and a label.
+
+You could compare a Symfony Form Theme to a PrestaShop theme: a PrestaShop theme 'skins' the Front office but does not hold its logic. It defines how the Front office looks but does not dictate how it behaves. The same principle can be followed for the Form Theme: it defines how a form looks but does not dictate its features.
+
+Diving deeper in details, in the Symfony Bootstrap Form Theme, file inputs are for example being rendered following this macro:
+{% raw %}
+```html
+{% block file_widget -%}
+    <{{ element|default('div') }} class="custom-file">
+        {%- set type = type|default('file') -%}
+        {{- block('form_widget_simple') -}}
+        {%- set label_attr = label_attr|merge({ class: (label_attr.class|default('') ~ ' custom-file-label')|trim })|filter((value, key) => key != 'id') -%}
+        <label for="{{ form.vars.id }}" {% with { attr: label_attr } %}{{ block('attributes') }}{% endwith %}>
+            {%- if attr.placeholder is defined and attr.placeholder is not none -%}
+                {{- translation_domain is same as(false) ? attr.placeholder : attr.placeholder|trans({}, translation_domain) -}}
+            {%- endif -%}
+        </label>
+    </{{ element|default('div') }}>
+{% endblock %}
+```
+{% endraw %}
+
+The main benefit of this system is that, since 'what is my form structure' and 'how is my form rendered' are completely separated, you can modify one without altering the other.
```

Comment:
```suggestion
The main benefit of this system is that, since the form structure and its render method are completely separated, you can modify one without altering the other.
```